src/app.js

```diff
@@ -13,21 +13,16 @@ export function app(props, container) {
   return globalActions
 
   function repaint() {
-    if (props.view && !skipRender) {
-      requestAnimationFrame(render, (skipRender = !skipRender))
+    if (props.view && !lock) {
+      requestAnimationFrame(render, (lock = !lock))
     }
   }
 
-  function render() {
-    flush(
-      (root = patchElement(
-        container,
-        root,
-        node,
-        (node = props.view(globalState, globalActions)),
-        (skipRender = !skipRender)
-      ))
-    )
+  function render(next) {
+    lock = !lock
+    if ((next = props.view(globalState, globalActions)) && !lock) {
```

Comment:
Do we need `&& !lock` here? Unless I'm mistaken, `render` is only called from L17, which is already guarded by a lock. So we will never unlock until `render` is actually called. So `!lock` on L23 will always be true. 
...again: unless I'm mistaken. (it does happen occasionally ;) )

Reply:
In fact.. I suspect the whole `if` statement around the `next = props.view...` may be superfluous

Reply:
Just a suggestion: maybe `lock` could be named `lockRepaint` just to make it a little clearer what we are locking and unlocking?

Reply:
This is all new behavior. Before, updating the state in a view via actions was possible, but left the state out of sync with the rendered VNode. I'll explain what's going on now.

The lock starts false. Then we call repaint to schedule the first render and the lock is now true. fAF kicks in and calls render. We enter render and the first thing we do now is toggle the lock. Now it's false, like at the start. This is new business.

Now we call the view with your state and actions. The view computes the next VNode. During the process, if the view or a component calls an action, we update the state and _call repaint_ as usual. The twist is that now, the lock, which is false at this point, allow us to schedule a new render. rAF is async so we return immediately, [toggle the lock again](https://github.com/hyperapp/hyperapp/blob/4b5708d83924462e3a86021169ad3ed9da8459bc/src/app.js#L17) and eventually return from the view. We're here now. `&& !lock`.

We know the lock was false, so `&& !lock` checks if we are still false and then proceeds to patch the DOM with the new VNode. Otherwise, the view (or maybe an async action from above and beyond...) toggled the lock. Then we know that another render has been scheduled, so we skip patchElement and get out knowing that we'll be back soon. 😎

When rAF calls render again, the lock is true, so we're like at the start and the whole process  starts over.

@zaceno @SkaterDad @lukejacksonn @okwolf

Reply:
@JorgeBucaran Ah right. Although the first thing `render` does is unlock, then calling `props.view` may cause `update` -> `repaint` -> locked again. So before we go ahead and patch, we make sure that didn't happen. If it did happen: skip patching. Well be right back soon anyway. 

Thanks for the illuminating explanation! I withdraw my comment.

Reply:
@JorgeBucaran Thanks for the detailed explanation.

Can you enlighten me on why this doesn't cause an infinite cycle?  Are actions called from within the view somehow different than actions called in response to a user event, like clicks?

```js
    view(state, actions) {
      actions.toggle() // <-- THIS ONLY GETS RUN ONE TIME
      return h('div', null, [
        h('button', { onclick: actions.otherThing }, 'Click Me'),
        h('pre', null, JSON.stringify(state, null, 2))
      ])
    }
```

Reply:
I was struggling to get my head around this bit too :P perhaps that is a sign that it could be refactored for readability. I suspect this foo has some size/perf justification behind it but now that you mention it it looks like the assignment of `next` is going to happen every call regardless of the state of `lock`.

Are you suggesting:

```
function render(next) {
  lock = !lock
  next = props.view(globalState, globalActions)
  if(!lock) { ... }
}
```

Reply:
@SkaterDad I believe this is something to do with `requestAnimationFrame` but I can not for the life of me explain why. I just know if you remove rAF and call render directly then it breaks things.

Reply:
@SkaterDad It will totally cause an infinite loop. Actions called within the view are no different that actions called in resp to user events, etc. To avoid an infinite loop, you will need to add logic.

```js

app({
  view(state, actions) {
    if (state.name === "") {
      actions.populate() && return
      // You don't need to return here, but you could, because we don't call
      // patchElement when the VNode is undefined.
    }

    return h("div", null, [
      h("button", { onclick: actions.up }, "Click Me"),
      h("pre", null, JSON.stringify(state, null, 2))
    ])
  },
  state: {
    value: 0,
    name: ""
  },
  actions: {
    up(state) {
      return { value: state.value + 1 }
    },
    populate(state) {
      return { name: "Hyperapp" }
    }
  }
})
```

Reply:
@JorgeBucaran I agree, it should cause an infinite loop.  However, I when I cloned this branch and tried that code, I only saw the action getting called once.  If you can't reproduce, I'll give it another try this evening if I get time.

Reply:
@SkaterDad If you get time and do try it, please post the entire code so I can cop&paste to check. Thanks!

Reply:
@JorgeBucaran Just tried it again (clone, build), and now I do get an infinite loop.  Not sure what I did wrong the first time around!

Reply:
Maybe my test the other day was accidentally on 0.15.1.  This doesn't loop.
https://www.webpackbin.com/bins/-Kx9XsEEufXxqggvDDYz

Reply:
@SkaterDad Odd!

![default](https://user-images.githubusercontent.com/56996/31907514-016c98ac-b86f-11e7-99bf-b9d9c0bb9f2a.gif)